code-dot-org/code-dot-org
# add ruby script to migrate units without importing CB content

Continues work on https://codedotorg.atlassian.net/browse/PLAT-830. Most of the code in 0bbea892e538b2d75059398dbdc4d8a3a0d58e70 was copied from https://github.com/code-dot-org/code-dot-org/blob/463c5c603b2de5ed33587c6b6238825b129b87d2/bin/curriculum/import_unit_details.rb .

## Testing story

manually verified as follows:
* `valentine` unit overview page appears unchanged (no View Lesson Plan links)
* `ap-csa` unit overview page appears unchanged (same View Lesson Plan links)
* `AlgebraA` unit can be migrated without errors due to upper case letters in the name
* legacy teacher resources in CSF, CSD and CSP courses from previous years are protected, which I verified as follows
```
Dave-MBP:~/src/cdo (migrate-unit-script)$ bundle exec bin/curriculum/migrate_unit.rb -u coursea-2020 -v
loading rails environment...
rails environment loaded in 33 seconds.
found code studio script name coursea-2020 with id 663
bundler: failed to load command: bin/curriculum/migrate_unit.rb (bin/curriculum/migrate_unit.rb)
RuntimeError: unsafe to migrate launched script containing teacher resources
  bin/curriculum/migrate_unit.rb:71:in `block in main'
  ...
```

```
[development] dashboard > unsafe_scripts = Script.all.select {|script| !script.is_migrated && script.teacher_resources.present? && script.launched?}
[development] dashboard > unsafe_scripts.map(&:name)
=> ["coursea-2018", "coursea-2019", "coursea-2020", "courseb-2018", "courseb-2019", "courseb-2020", "coursec-2018", "coursec-2019", "coursec-2020", "coursed-2018", "coursed-2019", "coursed-2020", "coursee-2018", "coursee-2019", "coursee-2020", "coursef-2018", "coursef-2019", "coursef-2020", "csd1-2017", "csd1-2018", "csd1-2019", "csd1-2020", "csd2-2017", "csd2-2018", "csd2-2019", "csd2-2020", "csd3-2017", "csd3-2018", "csd3-2019", "csd3-2020", "csd3-virtual", "csd4-2017", "csd4-2018", "csd4-2019", "csd4-2020", "csd5-2017", "csd5-2018", "csd5-2019", "csd5-2020", "csd6-2017", "csd6-2018", "csd6-2019", "csd6-2020", "csp-create-2017", "csp-create-2018", "csp-create-2019", "csp-explore-2017", "csp-explore-2018", "csp-explore-2019", "csp1-2017", "csp1-2018", "csp1-2019", "csp1-2020", "csp10-2020", "csp2-2017", "csp2-2018", "csp2-2019", "csp2-2020", "csp3-2017", "csp3-2018", "csp3-2019", "csp3-2020", "csp3-research-mxghyt", "csp4-2017", "csp4-2018", "csp4-2019", "csp4-2020", "csp5-2017", "csp5-2018", "csp5-2019", "csp5-2020", "csp6-2020", "csp7-2020", "csp8-2020", "csp9-2020", "express-2018", "express-2019", "express-2020", "pre-express-2020"]
```

## Deployment strategy

waiting until follow-up work steps are done before serializing any large numbers of scripts on levelbuilder. also waiting for https://github.com/code-dot-org/code-dot-org/pull/42937 to stabilize the script json format.

## Follow-up work

* add an `--all` option to migrate all scripts except the `unsafe_scripts` listed above
* safely handle teacher_resources
